File: code/smv-debugger/src/main/java/org/fbme/smvDebugger/integration/SmvService.java
package org.fbme.smvDebugger.integration;

import org.fbme.smvDebugger.model.CounterexampleParser;
import java.nio.file.Path;
import java.util.Optional;
import org.fbme.smvDebugger.model.Counterexample;
import java.nio.file.Files;
import java.util.List;
import java.io.IOException;

public class SmvService {
  private final Fb2SmvService fb2SmvService;
  private final NuSmvService nuSmvService;
  private final NutracService nutracService;

  private final CounterexampleParser counterexampleParser;

  public SmvService(final ServicePathProvider provider) {
    fb2SmvService = new Fb2SmvService(provider.getFb2SmvServicePath());
    nuSmvService = new NuSmvService(provider.getNuSmvServicePath());
    nutracService = new NutracService(provider.getNutracServicePath());

    counterexampleParser = new CounterexampleParser();
  }

  public Optional<Counterexample> verify(final Path fbPath, final String specification) {
    try {
      final Path smvPath = fb2SmvService.convertFbToSmv(fbPath);

      final Optional<Path> rawCounterexamplePath = nuSmvService.getRawCounterexample(smvPath, specification);
      Files.delete(smvPath);

      if (rawCounterexamplePath.isEmpty()) {
        return Optional.empty();
      }

      final Path csvCounterexamplePath = nutracService.convertToCsv(rawCounterexamplePath.get());
      Files.delete(rawCounterexamplePath.get());

      final List<String> lines = Files.readAllLines(csvCounterexamplePath);
      Files.delete(csvCounterexamplePath);

      final Counterexample counterexample = counterexampleParser.parse(lines);
      return Optional.of(counterexample);
    } catch (final IOException e) {
      throw new RuntimeException(e);
    }
  }

  public Counterexample parseCounterexample(Path exportedPath) {
    try {
      final List<String> lines = Files.readAllLines(exportedPath);
      return counterexampleParser.parse(lines);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }
}

File: code/smv-debugger/src/main/kotlin/org/fbme/smvDebugger/integration/SmvService.kt
package org.fbme.smvDebugger.integration

import org.fbme.smvDebugger.model.Counterexample
import org.fbme.smvDebugger.model.CounterexampleParser
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import kotlin.io.path.pathString

class SmvService(provider: ServicePathProvider) {
    private val nuSmvService: NuSmvService
    private val counterexampleParser: CounterexampleParser
    fun verify(fbPath: Path, specification: String): Path? {
        return try {
            //val smvPath = fb2SmvService.convertFbToSmv(fbPath)
            val smvPath = fbPath.pathString.substring(0, fbPath.pathString.lastIndexOf(".")) + ".smv"

            val rawCounterexamplePath = nuSmvService.getRawCounterexample(File(smvPath).toPath(), specification)
        //    Files.delete(smvPath)
            if (rawCounterexamplePath.isEmpty) {
                return null
            }
            return rawCounterexamplePath.get()


        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    fun parseCounterexample(exportedPath: Path?): Counterexample? {
        return try {
            val lines = Files.readAllLines(exportedPath)
            counterexampleParser.parse(lines)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    init {
        nuSmvService = NuSmvService(provider.nuSmvServicePath)
        counterexampleParser = CounterexampleParser()
    }
}
